openprescribing/data/rxdb/query.py

```diff
@@ -0,0 +1,89 @@
+import numpy
+from scipy.sparse._sparsetools import coo_todense
+
+from openprescribing.data.utils.duckdb_utils import (
+    FLOAT_TYPES,
+    NUMERIC_TYPES,
+    UNSIGNED_INTEGER_TYPES,
+)
+
+
+# Sets the number of rows we fetch in each batch from DuckDB. There's no perfect answer
+# to what size these batches should be, but here are some considerations:
+#
+#  * They should probably be a multiple of 2048 as that's the natural DuckDB "data
+#    chunk" size (see https://duckdb.org/docs/stable/clients/c/data_chunk).
+#
+#  * If they are too small then we will spend too much time hopping back and forth
+#    between fast native code and Python and that will slow things down.
+#
+#  * If they are too large then we force DuckDB to consume lots of memory while building
+#    the batch and, observationally, this seems to be slower (maybe while we're
+#    processing each batch DuckDB is preparing the next one in parallel?).
+#
+# The current value was determined by some crude profiling and a simple binary search to
+# land on what looked like the optimal value. It may well be possible to improve it.
+RECORD_BATCH_SIZE = 2048 * 64
+
+
+def get_grouped_sum_array(cursor, row_count, col_count, sql, parameters=None):
```

Comment:
Naming: should this be `get_grouped_sum_matrix`?

Reply:
Yes, quite possibly. I did want to distinguish it from the `LabelledMatrix` type somehow and make it clear that it returned something more primitive. I guess it could be even more precise as `get_grouped_sum_ndarray()`.

Reply:
Mild preference for `get_grouped_sum_ndarray()`, but up to you.